snapshot/git/gitfiler/pool.go

```diff
@@ -0,0 +1,89 @@
+package gitfiler
+
+import (
+	"github.com/scootdev/scoot/snapshot/git/repo"
+)
+
+// repoAndError holds a repo and an error (just so we can pass it across a channel)
+type repoAndError struct {
+	repo *repo.Repository
+	err  error
+}
+
+// NewRepoPool creates a new RepoPool populated with existing repos and a getter that can get new ones
+func NewRepoPool(getter RepoGetter, repos []*repo.Repository, doneCh chan struct{}) *RepoPool {
+	freeList := make([]repoAndError, len(repos))
+	for i, v := range repos {
+		freeList[i] = repoAndError{repo: v}
+	}
+
+	p := &RepoPool{
+		getter:    getter,
+		releaseCh: make(chan repoAndError),
+		reserveCh: make(chan repoAndError),
+		doneCh:    doneCh,
+		freeList:  freeList,
+	}
+	go p.loop()
+	return p
+}
+
+// RepoPool lets clients Get a Repository for an operation, and then Release it when done
+// RepoPool can create a new Repository by using a supplied RepoGetter.
+// New repos can be added by the client by Release'ing a new Repository
+// Cf. sync's Pool
+type RepoPool struct {
```

Comment:
When the Pool loop begins, it will start up a Get that will happen in the background. So if you start up a worker, you'll see it happening in the background before any requests arrive.